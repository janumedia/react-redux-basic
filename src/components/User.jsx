import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

import {deleteUser} from '../actions/userActions';

import './User.scss';

const mapDispatchToProps = dispatch => {
    return {
        deleteUser: userID => dispatch(deleteUser(userID))
    };
};

class User extends React.Component {

    constructor(props) {
        super(props);

        this.state = {
            loaded: false
        }
    }

    componentDidMount() {
        this.timerID = setTimeout(() => {
            this.setState({loaded: true})
        }, 10)
    }

    componentWillUnmount() {
        clearTimeout(this.timerID);
    }

    //optimization only render new user
    shouldComponentUpdate(nextProps, nextState) {
        return (this.props === nextProps)
    }

    deleteUser(e) {
        this.props.deleteUser(this.props.id);
    }

    render() {
        const {name, role} = this.props;
        return (
            <div className={`User ${this.state.loaded ? '' : 'hidden'}`}>
                <div>
                    <p>Name: {name}</p>
                    <p>Role: {role}</p>
                </div>
                <div>
                    <button onClick={this.deleteUser.bind(this)}>Delete</button>
                </div>
            </div>
        );
    }
}

User.propTypes = {
    name: PropTypes.string,
    role: PropTypes.string
}

export default connect(null, mapDispatchToProps)(User);