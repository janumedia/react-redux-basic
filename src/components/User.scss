.User {
    background-color: #ccc;
    margin-top: 20px;
    transition: opacity 1s;
    opacity: 1;
    position: relative;
    &.hidden {
        opacity: 0;
    }
    &.deleting {
        opacity: 0.4;
    }
    div:first-child {
        p {
            padding: 10px;
            margin: 0;
            &:last-child {
                padding-top: 0;
            }
        }
    }
    div:last-child {
        position: absolute;
        right: 10px;
        top: 0;
        transform: translateY(50%);
        button {
            padding: 10px;
            text-transform: uppercase;
        }
    }
}