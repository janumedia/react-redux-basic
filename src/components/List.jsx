import React from 'react';
import PropTypes from 'prop-types';

import User from './User';

const List = ({users}) => {
    //sort by id reverse
    const listUsers = [...users].sort((userA, userB) => (userA.id > userB.id) ? -1 : 1)
                           .map(({name, role, id}) => <User name={name} role={role} id={id} key={id}/>);
    return (
        <ul>{listUsers}</ul>
    )
}

List.propTypes = {
    users: PropTypes.arrayOf(
        PropTypes.shape({
            id: PropTypes.string.isRequired,
            name: PropTypes.string.isRequired,
            role: PropTypes.string.isRequired
        })
    )
}

export default List;