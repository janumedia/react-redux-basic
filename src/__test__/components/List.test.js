import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

import List from '../../components/List';
import User from '../../components/User';

describe('List component', () => {
    
    const props = {
        users: [
            {
                id: 20,
                name: "Charlie Jackson",
                role: "CTO"
            },
            {
                id: 1,
                name: "John Doe",
                role: "CEO"
            },
            {
                id: 3,
                name: "Dave Esco",
                role: "Engineer"
            },
            {
                id: 2,
                name: "Charlie Jackson",
                role: "CTO"
            }
        ]
    };

    let component, totalUsers = props.users.length;
    
    beforeEach(() => {
        component = shallow(<List {...props} />);
    });

    it('should render properly', () => {
        expect(component.type()).toEqual('ul');
        expect(component.children()).toHaveLength(totalUsers);
        expect(component.childAt(0).type()).toEqual(User);
        expect(component.childAt(totalUsers-1).type()).toEqual(User);
        expect(toJson(component)).toMatchSnapshot();
    });

    it('should render items short by "id" descending', () => {
        const shortedUsers = props.users.sort((userA, userB) => userA.id > userB.id ? -1 : 1);
        const matches = component.children().map(node => node)
        .filter((node, index) => {
            return node.props().id == shortedUsers[index].id;
        }).length;

        expect(matches).toEqual(props.users.length);
    })
})