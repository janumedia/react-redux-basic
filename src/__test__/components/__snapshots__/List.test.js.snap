// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List component should render properly 1`] = `
<ul>
  <Connect(User)
    id={20}
    key="20"
    name="Charlie Jackson"
    role="CTO"
  />
  <Connect(User)
    id={3}
    key="3"
    name="Dave Esco"
    role="Engineer"
  />
  <Connect(User)
    id={2}
    key="2"
    name="Charlie Jackson"
    role="CTO"
  />
  <Connect(User)
    id={1}
    key="1"
    name="John Doe"
    role="CEO"
  />
</ul>
`;
